### PR TITLE
Install all plugins during bats tests

### PR DESCRIPTION
**Pull request by nik9000**
Related to #12717

### CODE REVIEW

**Comment by nik9000** [comment]
This only installs and removes that plugins using the plugin script - it doesn't try to run elasticsearch with them all installed. I'm not sure vagrant the vagrant tests are the appropriate place to do that since they are ostensibly repeated once per target OS.

**Comment by nik9000** [comment]
Ping @tlrx  for more fun bats code to review. This one has a symlink!

File: qa/vagrant/src/test/resources/packaging/scripts/25_tar_plugins.bats [hunk 1]
```diff
@@ -0,0 +1 @@
+plugin_test_cases.bash
```
Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/25_tar_plugins.bats]: This is a symlink but its hard to read in github's ui.

**Comment by nik9000** [comment]
@brwe , this is one of the in flight bats changes.

**Comment by nik9000** [comment]
@spinscale, want to review some packaging tests?

**Comment by drewr** [comment]
Needed to use `VAGRANT_DEFAULT_PROVIDER=virtualbox` to get this to work for me. I suggest we explicitly set `--provider=virtualbox` wherever we call `vagrant up`.

**Comment by nik9000** [comment]
> Needed to use VAGRANT_DEFAULT_PROVIDER=virtualbox to get this to work for me. I suggest we explicitly set --provider=virtualbox wherever we call vagrant up.

I thought that was the default. I can certainly do that though.

**Comment by dakrone** [comment]
> Needed to use VAGRANT_DEFAULT_PROVIDER=virtualbox to get this to work for me.

Same here, otherwise vagrant tries to use libvirt or something.

**Comment by nik9000** [comment]
> Same here, otherwise vagrant tries to use libvirt or something.

wat? The default for vagrant for forever was virtualbox. Anyway, #13217.

**Comment by drewr** [comment]
@nik9000 It should be the default, but somehow because I use lxc in other contexts it seems to be sticking around somewhere. I even ran with `env - mvn ....` and it found it. Maybe just Vagrant :hankey:.

**Comment by drewr** [comment]
`env VAGRANT_DEFAULT_PROVIDER=virtualbox mvn -Dtests.vagrant clean verify` built successfully :bowtie:

**Comment by nik9000** [comment]
I've added the 2.0.0 label to this to make it clear that this is going to the 2.0 branch. I'll need to merge #13223 before I can merge this to the 2.0 branch but I'll get it before closing this pr.

**Comment by nik9000** [comment]
@tlrx, would you mind reviewing this on its own? I think it might be ok and its worth getting in if we can.

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash [hunk 2]
```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env bats
+
+# This file is used to test the installation and removal
+# of plugins with a tar gz archive.
```
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: Not only tar gz archive now

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    }
+    export ESHOME=/tmp/elasticsearch
+    export_elasticsearch_paths
+else
+    if is_rpm; then
+        GROUP='RPM PLUGINS'
+    elif is_dpkg; then
+        GROUP='DEB PLUGINS'
+    fi
+    export ESHOME="/usr/share/elasticsearch"
```
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: That would have been great to use ${packaging.elasticsearch.home.dir} var here but I guess it's not possible

**Comment by nik9000** [comment]
> @tlrx, would you mind reviewing this on its own? I think it might be ok and its worth getting in if we can.

Thanks! I'll read these and fix.

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    # Clean up after the last time we ran this test
+    rm -rf /tmp/config.*
+
+    move_config
+
+    CONF_DIR="$ESCONFIG" install_jvm_example
+    CONF_DIR="$ESCONFIG" remove_jvm_example
+}
+
+@test "[$GROUP] install jvm-example plugin with a custom ES_JAVA_OPTS" {
```
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: There's good chance that this test becomes obsolete soon with https://github.com/elastic/elasticsearch/pull/12801... I'm +1 for removing it now

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+@test "[$GROUP] install azure plugin" {
+    install_and_remove_special_plugin cloud azure azure-core-*.jar
+}
+
+@test "[$GROUP] install gce plugin" {
+    install_and_remove_special_plugin cloud gce google-api-client-*.jar
+}
+
+@test "[$GROUP] install delete by query" {
+    install_plugin delete-by-query "$(readlink -m delete-by-query-*.zip)"
```
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: Why this special treatment?

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+@test "[$GROUP] install python plugin" {
+    install_and_remove_special_plugin lang python jython-standalone-*.jar
+}
+
+@test "[$GROUP] install size mapper" {
+    install_plugin mapper-size "$(readlink -m mapper-size-*.zip)"
+    remove_plugin mapper-size
+}
```
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: :heart_eyes: those tests

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash (hunk 3 above)
Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: I'm not really sure.... I don't think so?

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash (hunk 4 above)
Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: It looks like it just won't work with that in there. I can drop the test.

File: qa/vagrant/src/test/resources/packaging/scripts/plugins.bash [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+# under the License.
+
+# Install a plugin an run all the common post installation tests.
+install_plugin() {
+    local name=$1
+    local path="$2"
+
+    assert_file_exist "$path"
+
+    "$ESHOME/bin/plugin" install "file://$path"
```
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugins.bash]: Doc says `file:$path`... I don't really care but maybe it's nicer to test what is documented?

File: qa/vagrant/src/test/resources/packaging/scripts/plugins.bash [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    local relativePath=${1:-$(readlink -m jvm-example-*.zip)}
+    install_plugin jvm-example "$relativePath"
+
+    assert_file_exist "$ESHOME/bin/jvm-example"
+    assert_file_exist "$ESHOME/bin/jvm-example/test"
+    assert_file_exist "$ESCONFIG/jvm-example"
+    assert_file_exist "$ESCONFIG/jvm-example/example.yaml"
+
+    echo "Running jvm-example's bin script...."
+    "$ESHOME/bin/jvm-example/test" | grep test
```
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugins.bash]: Nice

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash (hunk 5 above)
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: OK Got it - please ignore my comment

Comment (nik9000) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: It didn't really have a "prefix". I bet I can get that working with the other. Let me see about that....

File: qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash (hunk 4 above)
Comment (tlrx) [inline qa/vagrant/src/test/resources/packaging/scripts/plugin_test_cases.bash]: ++

**Comment by tlrx** [comment]
I like these changes, great work @nik9000 ! I read carefully the bats files and I made some minor comments. 

Tests succeed on my laptop for precise, trusty and centos-7 (still have to remove the elasticsearch-rpm dependency to get them work). Please let me know if you want me to test a specific VM.

The last and most important thing I see is that for now the test installs a plugin, check for files and remove the plugin. That would be nice to start elasticsearch with the installed plugin and checks that it started correctly and that the plugin is referenced in Nodes Info API... What do you think?

**Comment by nik9000** [comment]
> Tests succeed on my laptop for precise, trusty and centos-7 (still have to remove the elasticsearch-rpm dependency to get them work). Please let me know if you want me to test a specific VM.

Those vms are fine.

> The last and most important thing I see is that for now the test installs a plugin, check for files and remove the plugin. That would be nice to start elasticsearch with the installed plugin and checks that it started correctly and that the plugin is referenced in Nodes Info API... What do you think?

Yeah - we probably should do it. I think that should be in another pr though.

**Comment by tlrx** [comment]
> Yeah - we probably should do it. I think that should be in another pr though.

Sounds good. Can you create an issue please?

**Comment by tlrx** [comment]
LGTM then

**Comment by nik9000** [comment]
> Yeah - we probably should do it. I think that should be in another pr though.

#13254



> LGTM then

In that case I'll squash, rebase, and merge.

**Comment by nik9000** [comment]
Merged to master. Tried to merge it to 2.0 and found some unbackported stuff:
391ea379e2799037944741b8c626702434e931c0
0b650ed203939364bd9133041b6939deaa72b9d7

Trying to backport those before merging this into 2.0.

**Comment by nik9000** [comment]
> Trying to backport those before merging this into 2.0.

Did that in #13255.

**Comment by nik9000** [comment]
And merged to 2.0! Yay! All done.

**Pull request merged**